### DISCUSSION TITLE
how to use nebulaLogger in namespace unmanaged packaged environment

### DISCUSSION THREAD

**Discussion by akhtarhm** [jongpie/NebulaLogger#297] [Q&A] [answered]
Hi @jongpie I have installed unlocked nebulalogger in my sandbox and added the nebula apex, lwc code snippets in my apex classes and components. My sandbox components are deployed to prod with unmanaged packaged using namespace. 

so question here is do i need to package the nebulalogger with same namespace that my production has or how it is? 
How should I use nebula code in my packaged code which has namespace, Please suggest.

Thank you!!

**Comment by akhtarhm** [comment]
Any suggestions @jongpie will be much appreciated.

**Comment by jongpie** [comment]
Hi @akhtarhm - apologies for the late response! Could you clarify some details on your situation? Is your namespaced package something that will only be used internally at your company? Or will the package be installed in multiple client orgs?

**Reply by akhtarhm** [reply]
My package will be used internally only. So basically i have to use nebulla logger in apex and lwc which comes under that package.

**Reply by jongpie** [reply]
Ok, in that case, I think you have 3 possible options:

### Option 1: Include Nebula Logger's Metadata in Your Unlocked Package
This might be the simplest option - for this to work, you would need to download the metadata for Nebula Logger, and include it within your package's directory. 

With this approach...
- ...any time you create a new package version for your package, it will include your own metadata + all of the metadata for Nebula Logger.
- ...if you later want to upgrade Nebula Logger, you will have to handle upgrading Nebula Logger yourself by downloading a new version of Nebula Logger's metadata --> incorporate the new version of Nebula Logger metadata into your package's directories --> create a new package version.

### Option 2: Use the Managed Package Version of Nebula Logger (with a namespace) as a Package Dependency
Since your package has a namespace, you cannot directly use Nebula Logger's unlocked package as a dependency because it does not have a namespace. There is a way to workaround for this (Option 3 below) if you want to use the unlocked (non-namespaced) version of Nebula Logger, but another option is declare Nebula Logger's managed package (that uses the namespace `Nebula`) as a dependency for your package. To configure this, you would update your `sfdx-project.json` to have something similar to:

```json
{
    "namespace": "My_Namespace",
    "sfdcLoginUrl": "https://login.salesforce.com",
    "sourceApiVersion": "54.0",
    "packageDirectories": [ {
            "package": "My Example Package",
            "path": "./force-app/",
            "versionName": "Beta Release",
            "versionNumber": "0.9.0.NEXT",
            "dependencies": [
                {
                    "Nebula Logger - Managed Package@4.7.0-spring-22-release"
                }
            ]           
        }
    ],
    "packageAliases": {
        "Nebula Logger - Managed Package@4.7.0-spring-22-release": "04t5Y0000015lXNQAY",
        "My Example Package": "0Ho5Y000000XYZABC0",
        "My Example Package@4.0.0-9-managed-package-release": "04t5Y000000QWERTYU"
    }
}
```

With this approach...
- ...you would have to install the managed package version of Nebula Logger before you can deploy/install your own metadata - this includes scratch orgs, sandboxes, and production.
- ...there are [some limitations to be aware of](https://github.com/jongpie/NebulaLogger/wiki/Unlocked-vs-Managed-Package) with using Nebula Logger's managed package. Some features simply do not work as well when using the managed package.
- ...if you later want to upgrade Nebula Logger, you could do so independently from upgrading your own namespaced package

### Option 3: Use the Unlocked Package Version of Nebula Logger (with no namespace) as a Package Dependency
This is the most complex option to setup, but provides some benefits. However, I have not fully tested out this approach, so I can't say exactly what steps would need to be taken - if you're interested in this option, let me know and I can try to test it out in a next few weeks.

I hope this helps, but let me know if you have any follow-up questions!

**Reply by akhtarhm** [reply]
Thanks for reply @jongpie, will try the options-1. will let you know if have any problem

**Reply by akhtarhm** [reply]
i tried the option 2 but it's not allowing me to update sfdx-project.json with the above content.
Getting below errors.
![image](https://user-images.githubusercontent.com/81279920/173528870-994e8e0e-9d90-4564-8262-efce20ef50f0.png)

and if we able to update sfdx-project.json, do we need to move this change version of sfdx-project.json to prod environment also?

@jongpie

**Reply by akhtarhm** [reply]
I am trying to use packaged version but getting errors:
![image](https://user-images.githubusercontent.com/81279920/173530560-4741d22a-f222-43ba-a8a7-da4558e6c225.png)


@jongpie

**Reply by jongpie** [reply]
Hi @akhtarhm - a couple of follow up questions for you:

1. I was under the impression that you are building your own package, and that your own package has a namespace - but based on your first screenshot of `sfdx-project.json`, it seems as though you don't have a namespace set, nor does it look like you're building a package. Can you clarify on your scenario? Are you simply deploying metadata to your org? Or are you actually creating package versions & installing those in your orgs?
2. For your second screenshot that says "No MODULE named Nebula:logger found", do you have the Nebula Logger managed package installed in the target org? You'll need to have the managed package already installed in the org before you deploy/install your own metadata.

**Reply by akhtarhm** [reply]
Thanks for reply @jongpie. My reply:
1.  The 1st screenshot is from my dev sb where i am trying to add nebula code in my LWC. Yes we do packaging but not in dev sb, we have separate packaging environment against which our PR validated. Basic flow goes like: Dev sb ---> Packaging environment ---> QA environment ---> Production environment. I have installed the nebula package in my Dev sb, trying to add nebula LWC code in my components so that i can push those changes to Packaging environment where we will build our own package with namespace DeltakSRP. Once packaging completed, we will install our own package + nebula package in QA environment and so on.

2. Yes, I do have manage nebula with in my target environment where i am trying to deploy.  

![image](https://user-images.githubusercontent.com/81279920/174223396-8ed4d701-3646-4abe-8495-1267bdc3a9d3.png)


Please suggest how I should process

**Reply by akhtarhm** [reply]
As per below documentation, we can't use lwc from another namespace in our custom component:
1. https://ideas.salesforce.com/s/idea/a0B8W00000GdXQ3UAN/use-a-lightning-web-component-from-another-packagenamespace
2. https://salesforce.stackexchange.com/questions/301980/not-being-able-to-reference-packaged-custom-lwcs-from-other-lwcs-outside-the-pac?rq=1

**Reply by jongpie** [reply, marked as answer]
@akhtarhm this is now [available in `v4.13.0`](https://github.com/jongpie/NebulaLogger/releases/tag/v4.13.0)
